src/vehicle.cpp

```diff
@@ -5704,190 +5705,180 @@ bool vehicle::shift_if_needed() {
+            // Ignore the frame being destroyed
+            if( parts_in_square[index] == p ) {
+                continue;
+            }
+
+            if( parts[parts_in_square[index]].hp == 0 ) {
+                // Tearing off a broken part - break it up
+                if( g->u.sees( pos ) ) {
+                    add_msg(m_bad, _("The %s's %s breaks into pieces!"), name.c_str(),
+                            parts[ parts_in_square[ index ] ].name().c_str() );
                 }
+                break_part_into_pieces(parts_in_square[index], pos.x, pos.y, true);
             } else {
-                //Just break it off
-                if(g->u.sees( pos )) {
-                    add_msg(m_bad, _("The %1$s's %2$s is destroyed!"),
-                                    name.c_str(), parts[ p ].name().c_str() );
+                // Intact (but possibly damaged) part - remove it in one piece
+                if( g->u.sees( pos ) ) {
+                    add_msg(m_bad, _("The %1$s's %2$s is torn off!"), name.c_str(),
+                            parts[ parts_in_square[ index ] ].name().c_str() );
                 }
-                break_part_into_pieces(p, pos.x, pos.y, true);
-                remove_part(p);
+                item part_as_item = parts[parts_in_square[index]].properties_to_item();
+                g->m.add_item_or_charges( pos, part_as_item, true );
             }
+            remove_part( parts_in_square[index] );
+        }
+        /* After clearing the frame, remove it if normally legal to do
+         * so (it's not holding the vehicle together). At a later date,
+         * some more complicated system (such as actually making two
+         * vehicles from the split parts) would be ideal. */
+        if( can_unmount(p) ) {
+            if( g->u.sees( pos ) ) {
+                add_msg(m_bad, _("The %1$s's %2$s is destroyed!"),
+                        name.c_str(), parts[ p ].name().c_str() );
+            }
+            break_part_into_pieces( p, pos.x, pos.y, true );
+            remove_part(p);
         }
-        return dmg;
+    } else {
+        //Just break it off
+        if( g->u.sees( pos ) ) {
+            add_msg(m_bad, _("The %1$s's %2$s is destroyed!"),
+                            name.c_str(), parts[ p ].name().c_str() );
+        }
+
+        break_part_into_pieces( p, pos.x, pos.y, true );
+        remove_part( p );
     }
 
-    int tsh = part_info(p).durability / 10;
-    if (tsh > 20) {
-        tsh = 20;
+    return dmg;
+}
+
+bool vehicle::explode_fuel( int p, damage_type type )
+{
+    const itype_id &ft = part_info(p).fuel_type;
+    struct fuel_explosion {
+        // TODO: Move the values below to jsons
+        int explosion_chance_hot ;
+        int explosion_chance_cold;
+        float explosion_factor;
+        bool fiery_explosion;
+        float fuel_size_factor;
+    };
+
+    static const std::map<itype_id, fuel_explosion> explosive_fuels = {{
```

Comment:
subjectively this should be in JSON but this might be more than you want to tackle in this PR?